### DISCUSSION TITLE
Typing issue on selectinload() expected "Literal['*'] | QueryableAttribute[Any]"

### DISCUSSION THREAD

**Discussion by copdips** [fastapi/sqlmodel#871] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
query = select(User).options(selectinload(User.team)).offset(offset).limit(limit)
```

### Description

Hello,

I'm running a one-to-many,  Team to User classic relationship.

Table User declares field `team_id` which is a FK to `Team.id`. and relationship `team` for loading.

```python
class User(SQLModel, table=True)
    ...
    team_id: str | None = Field(
        default=None,
        foreign_key="team.id",
    }
    team: Team | None = Relationship(back_populates="users")
```

All is good, except a MyPy typing error on `selectinload()` method:

```console
error: Argument 1 to "selectinload" has incompatible type "Team | None"; 
expected "Literal['*'] | QueryableAttribute[Any]"  [arg-type]
```

When using: 

```python
from sqlalchemy.orm import selectinload

query = select(User).options(selectinload(User.team)).offset(offset).limit(limit)
```

### Operating System

Linux

### SQLModel Version

0.0.16
SQLAlchemy           2.0.29

### Python Version

Python 3.11.7

**Comment by AlPashy** [comment]
Have you solved it?, cause i get the same exception

**Reply by copdips** [reply]
sorry nope

**Comment by itsbekas** [comment]
Hey 🙂 As a note for the future, please try to provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Following the [Discussion Guidelines](https://github.com/tiangolo/sqlmodel/discussions/new?category=questions) is important to help us help you.

This is my attempt at recreating your situation:

```py
from sqlmodel import Field, Relationship, SQLModel, create_engine, Session, select
from sqlalchemy.orm import selectinload

class Team(SQLModel, table=True):
    id: int = Field(primary_key=True)
    name: str = Field()
    users: list["User"] = Relationship(back_populates="team")

class User(SQLModel, table=True):
    name: str = Field(primary_key=True)
    team_id: str | None = Field(
        default=None,
        foreign_key="team.id",
    )
    team: Team | None = Relationship(back_populates="users")

engine = create_engine("sqlite:///example.db")
SQLModel.metadata.create_all(engine)

team1 = Team(name="Team Rocket")

user1 = User(name="Jessie", team=team1)
user2 = User(name="James", team=team1)

with Session(engine) as session:
    session.add(team1)
    session.add(user1)
    session.commit()

    query = select(User).options(selectinload(User.team)).offset(1).limit(1)

    users = session.exec(query)

    print([r for r in users])
```

However, this version runs as-is without any issues (I used SQLModel 0.0.16 and SQLAlchemy 2.0.29, but I'm using Python 3.12, though I doubt that's the cause). If your use-case is different, feel free to modify this example to better illustrate it 🙂 Maybe you can show us yours as well @AlPashy

**Reply by AlPashy** [reply]
I have  implemented  the recreation from you and mypy says:
load.py:30: error: Argument 1 to "selectinload" has incompatible type "Team | None"; expected "Literal['*'] | QueryableAttribute[Any]"  [arg-type]

newest sqlalchemy, sqlmodel version with python 3.11. Is there a alternative to solve this?

**Reply by copdips** [reply]
I tested in 3.12.2 with new venv (pip install sqlmodel mypy), same error

**Reply by itsbekas** [reply, marked as answer]
You're both right, I apologize. For some reason mypy wasn't picking up that error on my side.

In the meantime, I think I found the issue:
By changing `team: Team | None` to `team: Mapped[Team | None]` (note that `Mapped` is imported from `sqlalchemy.orm`), the error seems to disappear, while the example still works as intended. However, I'm not sure if this has any side effects, namely with Pydantic (assuming you're using FastAPI), so you might want to check that, just to be sure.

**Reply by copdips** [reply]
great, no more mypy errors, also tested with my fastapi, all good, thanks a lot for this solution.

**Reply by mbyrne00** [reply]
The problem with using `Mapped[Team]` I now get typing warnings when I try to access any property on the team (I left out the `| None` for now for simplicity).

```python
user.team.name
```

Warning ..... `Unresolved attribute reference 'name' for class 'Mapped'`


I'm not sure the generics are working properly. At least this is happening in pyCharm.


My solution was to leave the model definition alone, without any `Mapped[]` wrapper and change the type of the field when we use it in the join. I don't understand it completely but this is what made it happy, taking inspiration from the `col()` function in SQLModel.

```python
from sqlalchemy.orm import InstrumentedAttribute


_T = TypeVar("_T")


def attr(instance: _T) -> InstrumentedAttribute[_T]:
    return instance
```

Then in your join code, wrap `User.team` inside `attr()` ...

```python
query = select(User).options(selectinload(attr(User.team))).offset(1).limit(1)
```

**Reply by chriscarrollsmith** [reply]
That doesn't quite work, but you can do this:

```python
from sqlalchemy.orm import InstrumentedAttribute

query = select(User).options(selectinload(cast(InstrumentedAttribute, User.team))).offset(1).limit(1)
```
